### DISCUSSION TITLE
Integration of Mellinger Controller Library for Waypoint-based Trajectory Planning and Position Correction in Crazyflie Firmware

### DISCUSSION THREAD

**Discussion by amitx64** [bitcraze/discussions#1147] [Q&A]
Hello Crazyflie community,

I hope this message finds you well. I am currently working on a project that involves utilizing the Mellinger Controller Library provided in the Crazyflie firmware. My goal is to implement waypoint-based trajectory planning, where the Crazyflie will store target position coordinates as waypoints, similar to GPS waypoints, and execute an autonomous trajectory using the Mellinger controller. Additionally, I aim to incorporate position correction between waypoints until the next set of coordinates is received.

I would greatly appreciate any guidance, insights, or code snippets from the community regarding the integration of the Mellinger Controller Library for this purpose. If anyone has experience or suggestions on how to efficiently achieve waypoint storage, trajectory planning, and position correction using the Mellinger controller, your input would be invaluable.

Thank you in advance for your time and assistance!

**Comment by xkhainguyen** [comment]
Hi there,

Not sure if this is still of your interest but you can take a look at the [TinyMPC-integrated firmware](https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware/tree/main) to see how we store the full trajectory [waypoints](https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware/blob/main/examples/controller_tinympc_state_soc/src/landing_trajectory.hpp) and use MPC to execute it autonomously while considering several constraints.
Of course you can switch to Mellinger controller as well, all supported within our example.

**Reply by amitx64** [reply]
@xkhainguyen, I will try this. Thank you for sharing this.

**Reply by amitx64** [reply]
@xkhainguyen can I use `TinyMPC` with `LPS System`?

**Reply by xkhainguyen** [reply]
Yes, it is just the controller. You can use whatever localization methods you want: Flow deck, LPS, MoCap, etc
